File: 1382-balance-a-binary-search-tree/1382-balance-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def balanceBST(self, root: TreeNode) -> TreeNode:
        
        def preorder(node):
            if(not node):
                return 
            
            preorder(node.left)
            ordered.append(node.val)
            preorder(node.right)
        
        def construct(start, end):
            if(start > end):
                return None
            
            mid = (start + end) // 2
            print(ordered[mid])
            node = TreeNode(ordered[mid])
            node.left = construct(start, mid - 1)
            node.right = construct(mid + 1, end)
            return node
        
        ordered = []
        preorder(root)
        return construct(0, len(ordered) - 1)
        

File: 1382-balance-a-binary-search-tree/1382-balance-a-binary-search-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func balanceBST(root *TreeNode) *TreeNode {
    nodeVals := treeToList(root)
    return formTree(nodeVals)
}

func treeToList(root *TreeNode) []int {
    if root == nil {
        return []int{}
    }
    nodes := treeToList(root.Left)
    nodes = append(nodes, root.Val)
    nodes = append(nodes, treeToList(root.Right)...)
    return nodes
}

func formTree(vals []int) *TreeNode {
    size := len(vals)
    if size == 0 {
        return nil
    }
    mid := size / 2
    return &TreeNode {
        vals[mid],
        formTree(vals[:mid]),
        formTree(vals[mid + 1:]),
    }
}